Load improvements

gokmen opened the pull request:
There are some multiple fetches and duplicated requests on the example load for StackEditor (new one) following needs to be fixed or moved into payload as much as possible.

```
reqs on /Stack-Editor/58b5d77d0ca04a52ddd7026f
----------------------------------------------

-- took 4  >> JCustomPartials.some         -> do we still need this?
-- took 3  >> JName.one                    -> add to payload
-- took 9  >> JAccount.fetchAppStorage     -> add to payload
-- took 14 >> JUser.fetchUser              -> add to payload --------------
-- took 17 >> JStackTemplate.one           // application request         |
-- took 20 >> JComputeStack.some           -> not required                |
-- took 28 >> JCredential.some             -> add to payload              |
-- took 31 >> JCredential.some             -> shouldn't happen twice      |
-- took 2  >> JGroup.one                   -> ?? -------------------------+---
-- took 11 >> JMachine.some                -> not required                |  |
-- took 30 >> SocialChannel.byId           -> add to payload              v  |
-- took 8  >> JAccount.fetchEmail          -> ?? --------------------------  v
-- took 8  >> JGroup.one                   -> ?? -----------------------------
-- took 14 >> JComputeStack.checkRevision  -> add to payload
-- took 14 >> JComputeStack.checkRevision  -> shouldn't happen twice
-- took 9  >> Sidebar.fetchEnvironment     -> ?? already added to payload
-- took 11 >> Sidebar.fetchEnvironment     -> shouldn't happen twice
```

This is still wip maybe won't end up in the production but will update here with my findings.

gokmen committed:
ComputeController::reset will be deprecated

gokmen committed:
Expressify: log all request over remote

gokmen committed:
EnvironmentFlux: disable fetchers for now

gokmen committed:
ComputeStorage: initial commit

gokmen committed:
ComputeController: use computeStorage instead

gokmen committed:
Compute: remove deprecated helpers

gokmen committed:
Farewell JWorkspace

gokmen committed:
Remove JWorkspace related views tests and assets from application

gokmen committed:
EnvironmentDataProvider, ShareManager: remove workspace related flows

gokmen committed:
Environment: Flux: remove workspace related stores & tests

gokmen committed:
Environment: Flux: remove workspace related actions & actionTypes

gokmen committed:
Environment: Flux: remove workspace related getters

gokmen committed:
Update utils for workspace removals, remove ws related action triggers

gokmen committed:
ComputeController: remove domain management

gokmen committed:
ComputeController: update load flow, remove workspace related flows

gokmen committed:
Fix linter issues

gokmen committed:
Application: cleanup for deprecated parts

gokmen committed:
Disable ComputeController.reset calls until next iteration

gokmen committed:
Remote: fetchName: no need to re-fetch group instance

gokmen committed:
Payload: provide userEmail on payload

gokmen committed:
Payload: provide group socialApiChannel on payload

gokmen committed:
Payload: Sidebar: remove environmentData from payload

gokmen committed:
Compute: use new computeStorage instead

gokmen committed:
OnBoardingController: disable CustomPartials

gokmen committed:
ComputeController: use stack.machines instead

gokmen committed:
Account.fetchCombinedStorage: return if found in cache

gokmen committed:
IDE: Collab: disable workspace related requirements

gokmen committed:
IDE: routes from scratch

gokmen committed:
Payload: provide userStacks on payload along with machines

gokmen committed:
ComputeStorage: update getter to support queries for unique keys

gokmen committed:
ComputeStorage: use map instead

gokmen committed:
ComputeStorage: create initial schema of storage

gokmen committed:
ComputeController: checkGroupStacks: no need to re-fetch group

gokmen committed:
ComputeController: fetchStackTemplate{,s} with caching on storage

gokmen committed:
EnvironmentActions: loadStackTemplates over compute controller

gokmen committed:
getKdCmd: as app util: get kd install line

gokmen committed:
ResourceStateModal: no need to generate kd install line on load

gokmen committed:
ResourceStateModal: generate kd install line on the fly when needed

gokmen committed:
ComputeStorage: filter out undefined results from query results

gokmen committed:
CheckCredentialSteps: no need to check it if it’s completed once

gokmen committed:
CredentialForm: use onCreateNew helper instead

gokmen commented:
Loading on an IDE route was doing following additional calls to `remote.api` in order:
```
>> JCustomPartials.some        -- took 1
>> JName.one                   -- took 2
>> JAccount.fetchAppStorage    -- took 7
>> JStackTemplate.some         -- took 8
>> JComputeStack.some          -- took 13
>> JUser.fetchUser             -- took 14
>> JStackTemplate.some         -- took 17
>> JCredential.some            -- took 26
>> JGroup.one                  -- took 3
>> JMachine.some               -- took 10
>> JAccount.fetchEmail         -- took 7
>> JGroup.one                  -- took 7
>> JComputeStack.checkRevision -- took 10
>> SocialChannel.byId          -- took 18
>> JComputeStack.checkRevision -- took 14
>> JAccount.fetchAppStorage    -- took 6
>> JStackTemplate.one          -- took 7
>> JAccount.fetchOtaToken      -- took 4
>> JCredential.some            -- took 12
```

But now on this branch it only does following;

```
>> JStackTemplate.some         -- took 11
>> JStackTemplate.one          -- took 8
>> JCredential.some            -- took 15
```

gokmen committed:
ApplicationManager: getInstance use query supported helper instead

gokmen committed:
RemoteExtensions: inject extensions just after bongoInstace create

gokmen committed:
KodingKontrol: wait for maincontroller to be ready

gokmen committed:
JMachine.State: use State from JMachine or as inline string

In various places we need to have JMachine.States which are only
available once remote.api is ready. But we need them sync all the way
long. So, for now I’ve used them as string in such places ~ GG

gokmen committed:
Remote-Extension: JMachine support whole feature set of Machine

gokmen committed:
Environment: no need for Machine instance anymore JMachine’s are enough!

From now on we don’t need to create a new Machine instance from JMachine
New JMachine instances will have the same featureset of Machine instance
So, just fetching the JMachine instance will work as is.

gokmen committed:
Remove remaining Machine usages

gokmen committed:
Remove DummyMachine and use new JMachine instance with dummy data

gokmen committed:
ComputeStorage: query: support full data path on key if possible

gokmen committed:
ComputeStorage: set read-only if group is disabled

gokmen committed:
ComputeController: no more fetchStacks

gokmen committed:
EnvironmentActions: reinitStack -> checkTeamStack better naming

gokmen committed:
ComputeStorage: declarative storage data for known/supported models

gokmen committed:
JMachine::one: unify to support query string instead

gokmen committed:
JComputeStack: revive: update and return revived data for ::one$

gokmen committed:
JGroup: fetchGroupResources: pass client for revive

gokmen committed:
IDE: routes: use slug instead of label

gokmen committed:
Environment: filter machines per store on client side

From now on we only have machines list from backend and separating of
machines will be handled in client side in store definitions.

gokmen committed:
{Private, Team}TemplateStore: better updater for stackTemplate.title

gokmen committed:
ComputeStorage: revamp: declarative handlers and push/pop flows

gokmen committed:
Compute: use new storage push/pop flow

gokmen committed:
JStackTemplate: generateStack: update stores when needed

gokmen committed:
Environment: on StackRevisionChanged no need to re-fetch machines

gokmen committed:
Environment: getters: filter first then make the assignment

gokmen committed:
ComputeController: createDefaultStack: support options

gokmen committed:
JMachine: provide ready method

gokmen committed:
Stack-Editor: use debounce on title change

gokmen committed:
GetMachineOwner: no need to revive each machine again and again

It was the main reason of the most of the lags :\

gokmen committed:
Sidebar: generateStack: handle error in place where the call executed

gokmen committed:
ComputeController: fetchStackTemplate: use storage fetcher instead

gokmen committed:
Rename isClonedTemplate to fetchOriginTemplate and use computestorage

gokmen committed:
ComputeController: checkStackRevision: check only provided template

Currently this method was checking all the stack templates present in
stores on every change of any stacktemplate which was not necessary at
all. After this change this will still support to check all at once but
if a stackTemplateId provided it will only checked the changes on that

gokmen committed:
ComputeStorage: no need to revive already revived instances

gokmen committed:
ComputeStorage: only dispatch updated stack and it’s machines

gokmen committed:
ComputeController: reloadIDE: fix reload issue for the same path

gokmen committed:
SidebarMachineListItem: use states for event percentage

gokmen committed:
Environment: remove machine data bindings

gokmen committed:
Environment: no need to wait for computeController.ready

gokmen committed:
ComputeController: fix triggerRevive flow, remove not necessary calls

gokmen committed:
ComputeStorage: on push update the existing data if possible

gokmen committed:
IDE: code tidy up, more debug points and updates on KlientEventManager

gokmen committed:
Environment: code tidy up, return missing promises to keep chains safe

gokmen committed:
JStackTemplate.update: update templates storage on each update

gokmen committed:
ComputeStorage: update: extend existing instance on update

gokmen committed:
ComputeStorage: push: execute modifiers, prePush and postPush always

gokmen committed:
ComputeStorage: get/query: faster and simpler getters

gokmen committed:
ResourceStateModal: fix stack build flow by using updated data in store

usirin commented on client/home/lib/stacks/homestacksimport.coffee:
```diff
@@ -156,6 +156,6 @@ module.exports = class HomeStacksImport extends kd.CustomHTMLView
 
       unless err
-        EnvironmentFlux.actions.loadPrivateStackTemplates()
+        EnvironmentFlux.actions.loadStackTemplates()
```
I think there is a typo here, there is no `loadStackTemplates` action defined.

gokmen committed:
Environment: better naming for loadStackTemplates

gokmen committed:
ApplicationManager: getInstance: faster :car:

gokmen commented on client/app/lib/util/fetchorigintemplate.coffee:
```diff
@@ -0,0 +1,11 @@
+  return callback null  unless stackTemplate
+
+  originalStackTemplateId = stackTemplate.config?.clonedFrom
+  return callback null  unless originalStackTemplateId
```
on all of the remote operations we're providing the following structure on callbacks; `(err, data)` `err` can be anything about the operation itself and the `data` is the result. if an operation not produce any data it doesn't mean that there is an error. so, in application wise;

```coffee

remote.api.JSession.activeSession (err, session) ->
  if err then handleError err
  else if not session then handleNoSession
  else handleSession session
```
would be the simpler example.

gokmen committed:
ComputeProvider: fix createGroupStack flow by using ComputeStorage

gokmen committed:
Debug: add more debug points

gokmen committed:
Scriptblock: fetch or create the appstorage for new users

gokmen committed:
Flux: Search: use allMachines getter instead

gokmen committed:
MachineStore: no need for a seperate getMachineOwner util

gokmen committed:
Stores: fix own machines store

gokmen committed:
IDE: Routes: if found by slug use the one that user owns

Since machine labels/slugs are unique under user scope, if we need to
solve a machine by slug we can only do that by checking out it’s
ownership. For the rest of the machines uid will provide global unique
identifier.

usirin committed:
sidebar: make async calls after component mounted

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: change props of no stacks widget

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: change state key for all templates

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: pass correct props to sidebar no stacks

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: don't use team from flux store

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/environment: remove SharedMachinesStore

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/environment: remove OwnMachinesStore

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/environment: remove CollaborationMachinesStore

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/environment: set test machine's type to 'own'

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/env: handle reject invitation in MachinesStore

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/environment: listen revive compute ctrl revive events

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: listen stack.machine events for each item

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

gokmen committed:
ProfileLinkContainer: fix React warning about not supported props

gokmen committed:
JMachine::setChannelId: to keep collaboration channelID on JMachine

gokmen committed:
JMachine::setChannelId tests and swagger definition update

gokmen committed:
Collaboration: over channelId on JMachine instances

gokmen committed:
Collaboration: use {get, set}ChannelId where needed for channels

gokmen committed:
SidebarMachineListItem: not all machines have stacks

gokmen committed:
Environment: Actions: approve/reject invitation flows

gokmen committed:
KodingKite: try to renew token when session is closed, add debug data

gokmen committed:
JMachine:channelId: change index to sparse

gokmen committed:
Collaboration: re-fetch updated machine instance after modifying it

gokmen committed:
NotificationController: handle machinelistupdated for guests

gokmen committed:
ComputeStorage: on pop support custom keys, on push ignore null data

gokmen committed:
CollaborationController: use machine instance from storage all the time

gokmen committed:
IDE: handle collaboration flows correctly, remove old way with ws’s

gokmen committed:
IDE: routes: handle collaboration only for shared machines

gokmen committed:
Collaboration: cleanup for workspace related codes

gokmen committed:
Collaboration: try to deny machine no matter what on leave session

gokmen committed:
IDE: routes: simplify the route for collaboration and fix linter issue

gokmen committed:
Intercom: disable intercom widget in the application

gokmen committed:
Environment: unify INVITATION_REJECTED action and action type

gokmen committed:
Machine: deny: pop the machine from storage once it’s denied

gokmen committed:
Collaboration: on leave session deny machine if it’s not permanent

Otherwise it reloads the current IDE to restart it without session

usirin committed:
sidebar/machineitem: add debug statements

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
remote-ext: add setApproved method to machine

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/env: fix acceptInvitation handler in store

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
flux/env: update cc storage on invitation accept

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

gokmen committed:
ComputeStorage: create helpers over stores for all related ones

gokmen committed:
ComputeStorage: use machines store where possible

gokmen committed:
ComputeStorage: use stacks store where possible

gokmen committed:
ComputeStorage: use templates store where possible

usirin committed:
flux/env: use machine from cc storage, don't revive

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
remote-ext: update cc storage on machine.approve

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
remote-ext: assign correct attr in machine.setApproved

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

gokmen committed:
ComputeController: destroyStack: handle IDE quit flow correctly

Before it was quiting from all IDE instances no matter what when a
stack is destroyed, with this change it will only quit from the ones
which has a mounted machine from destroyed stack, if not it will remain
as is.

usirin committed:
sidebar: don't bind to requiredInvitationMachine events

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: listen invitation events on individual machines

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: set inv machine id to null when clicked outside

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
sidebar: set leaving machine id to null when clicked outside

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

gokmen committed:
GroupsController: use channel.byId to subscribe realtime notifications

gokmen committed:
ComputeController: on StackTemplateChanged only check for changed one

sinan commented on client/app/lib/redux/dispatchInitialActions.coffee:
```diff
@@ -59,7 +59,7 @@ module.exports = dispatchInitialActions = (store) ->
   promise = loadAccount(store)
     .then -> loadGroup(store)
-    .then -> loadUserDetails(store)
+  # .then -> loadUserDetails(store)
```
check with @usirin

gokmen replied:
@usirin you were going to delete this requirement for payment flow afair, can u chime in?

sinan commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -17,6 +17,13 @@ module.exports = (options = {}) ->
+    if method.constructorName
+      console.log ">> #{method.constructorName}.#{method.method}"
+      console.log "   on #{method.id}"  if method.type is 'instance'
+      console.log "-- took #{duration}"
```
necessary?

gokmen replied:
it was for testing purposes, will be removed 👍

sinan commented on workers/social/lib/social/render/scriptblock.coffee:
```diff
@@ -5,20 +5,21 @@ module.exports = (options = {}, callback) ->
-  userWorkspaces      = null
-  userEnvironmentData = null
+  userStacks          = null
+# userEnvironmentData = null
```
delete instead of commenting out

sinan commented on workers/social/lib/social/render/scriptblock.coffee:
```diff
@@ -118,38 +129,42 @@ module.exports = (options = {}, callback) ->
+    # (fin) ->
+    #   bongoModels.Sidebar.fetchEnvironment client, (err, data) ->
+    #     userEnvironmentData = data
+    #     fin()
```
again delete

sinan reviewed (approved):
💯 a few minor notes, nothing important.

gokmen commented on client/app/lib/components/sidebarmachineslistitem/index.coffee:
```diff
@@ -27,66 +25,86 @@ module.exports = class SidebarMachinesListItem extends React.Component
     coordinates = getBoundingClientReact @refs.sidebarMachinesListItem
+
+    debug 'setcoordinates', coordinates
     @setState { coordinates: coordinates }
```
cc/ @usirin
